StrongInject.Generator/Statement.cs

```diff
@@ -16,6 +16,11 @@ namespace StrongInject.Generator
     {
         public string DisposeActionsName { get; } = "disposeActions_" + VariableName;
     }
+    internal sealed record OwnedCreationStatement(
+        string VariableName,
+        OwnedSource Source,
+        string DependencyName,
+        ImmutableArray<Operation> OperationsToDispose) : Statement;
```

Comment:
I wonder if it would be simpler to lower this in the same way as a DelegateCreationStatement, just using a local method instead of a delegate?

Reply:
I have a slight preference for the current output, and I don't have the sense (yet?) that this would be simpler to generate, but I'm willing to implement as a local function so you can compare the generator complexity.

Reply:
I was just thinking conceptually, it's clearer that this is it's own scope, so when you look at the test output each method defines a scope.

Reply:
I'm not sure I know what you mean by having its own scope. If scope relates only to disposal then fine, but if scope also relates to creation, it gets complicated. The variables have to be declared exactly as though they were part of the containing scope.

Reply:
Well, maybe they don't and it's a failure of my imagination. What other resulting code can you picture that declares B and C variables differently from the current test source, and is valid for injecting `Func<Owned<T>>`?

Reply:
Basically what I'm considering is that variables that are used by the `Owned` can't be shared with variables that are used outside the `Owned`. Maybe writing this as a local function might make that distinction clearer? I'm definitely willing to bend hear if you think that's not the case.

Reply:
Meaning `Func<A>` with `record A(Owned<B> b, C c);` and B depends on C also?

Reply:
Precisely

Reply:
I mean, I'm not sure we talked about this. How many instances of `C` should there be if you do this? And when does each instance get disposed?

```cs
class A
{
    public A(Owned<C> c1, Owned<C> c2, C c3, C c4)
    {
        Console.WriteLine(ReferenceEquals(c1.Value, c2.Value));
        Console.WriteLine(ReferenceEquals(c2.Value, c3.Value));
        Console.WriteLine(ReferenceEquals(c3.Value, c4.Value));
    }
}
```

Reply:
I'd be just as happy to say that only `Func<Owned<C>>` is supported and `Owned<C>` is not supported, since I can't think of a single scenario the latter unblocks.

Reply:
This is still not obvious to me, but I'll assume that you mean `Owned<C>` (which is not `Func<Owned<C>>`) is passed with a different C, and `c3` and `c4` get the same C. Then we're in good shape since the code is already doing this.

Reply:
So, jumping back to your original question:

> Basically what I'm considering is that variables that are used by the `Owned` can't be shared with variables that are used outside the `Owned`. Maybe writing this as a local function might make that distinction clearer? I'm definitely willing to bend hear if you think that's not the case.

I don't understand what is being asked. Are you look at the most recently-pushed tests and use them as an example to demonstrate the difference in resulting generated source that you have in mind?

Reply:
That makes sense. Are you picturing that the local method would return `Owned<X>` and construct the dependencies needed for `X`?

Reply:
To be clear, I think this will make the implementation more complex, not less, but it will make the generator output easier to read which is good.

Reply:
So what I would suggest is that in this test example:

```csharp
        if (Disposed)
            throw new global::System.ObjectDisposedException(nameof(Container));
        global::C c_0_1;
        global::B b_0_0;
        global::StrongInject.Owned<global::B> owned_0_1;
        global::C c_0_2;
        global::A a_0_0;
        c_0_1 = new global::C();
        try
        {
            b_0_0 = new global::B(c: c_0_1);
            try
            {
                owned_0_1 = new global::StrongInject.Owned<global::B>(b_0_0, () =>
                {
                    ((global::System.IDisposable)b_0_0).Dispose();
                    ((global::System.IDisposable)c_0_1).Dispose();
                });
                c_0_2 = new global::C();
                try
                {
                    a_0_0 = new global::A(b: owned_0_1, c: c_0_2);
                }
                catch
                {
                    ((global::System.IDisposable)c_0_2).Dispose();
                    throw;
                }
            }
            catch
            {
                ((global::System.IDisposable)b_0_0).Dispose();
                throw;
            }
        }
        catch
        {
            ((global::System.IDisposable)c_0_1).Dispose();
            throw;
        }
```

We instead generate (definitely mistakes here, done by hand, but should get concept across):

```csharp
        if (Disposed)
            throw new global::System.ObjectDisposedException(nameof(Container));
        global::StrongInject.Owned<global::B> owned_0_1;
        global::C c_0_2;
        global::A a_0_0;
        Owned<B> CreateOwnedB() {
               global::C c_0_1;
               global::B b_0_0;
               global::StrongInject.Owned<global::B> owned_0_1;
               c_0_1 = new global::C();
               try
               {
                   b_0_0 = new global::B(c: c_0_1);
                   try
                   {
                         owned_0_1 = new global::StrongInject.Owned<global::B>(b_0_0, () =>
                        {
                               ((global::System.IDisposable)b_0_0).Dispose();
                               ((global::System.IDisposable)c_0_1).Dispose();
                       });
                   }
               }
               catch
               {
                    ((global::System.IDisposable)b_0_0).Dispose();
                   throw;
               }
               return owned_0_1;
        }
        owned_0_1 = CreateOwnedB();
        c_0_2 = new global::C();
        try
        {
              a_0_0 = new global::A(b: owned_0_1, c: c_0_2);
        }
        catch
        {
              ((global::System.IDisposable)c_0_2).Dispose();
              throw;
        }
```

Do you think that looks better?

Reply:
> To be clear, I think this will make the implementation more complex, not less, but it will make the generator output easier to read which is good.

I think it will add complexit yin some ways, but will also simplify in others. For example I don't think you'll have to special case the disposal of Owned dependencies anymore. I think you might have to update `InstanceSourcesScope.Enter` to enter a new scope at Owned, as otherwise you'll have naming conflicts I think, but I'm not sure.

Reply:
Yes, it is easier to understand. I'll get started on this! I'll probably squash when I backtrack unless you would like me to maintain the first approach in the commit history.

Reply:
> I'll probably squash when I backtrack unless you would like me to maintain the first approach in the commit history.

No problem at all. I might squash the whole thing when I merge anyway.

Reply:
An interesting question is going to be how to unify `CreateOwnedb_1_0` etc into `CreateOwnedB`, emitted just once.

Reply:
That's an interesting hack. Obviously id prefer not to, if only for performance, but if you see no other solution go for it.

Another complexity - the local function needs to be the same asychronousness as the method it's in.

Reply:
I didn't realize there was a serious performance aspect and thought you might prefer it, so never mind. I'll start out with the obvious approach which can result in duplicate `CreateOwned_b_1_0`-style local functions being generated, and then I'll look for a way to deduplicate them.

Reply:
Moving discussion to https://github.com/YairHalberstadt/stronginject/pull/143#issuecomment-898977510.